compiler/noirc_frontend/src/hir_def/expr.rs

```diff
@@ -208,7 +212,28 @@ impl HirMethodCallExpression {
         object_type: Type,
         location: Location,
         interner: &mut NodeInterner,
-    ) -> ((ExprId, HirIdent), HirCallExpression) {
+    ) -> ((ExprId, HirIdent), HirCallExpression, usize) {
+        // Attach any generics which may not be a part of the function call's explicit generics
+        // This includes types such as implicit generics if the function is part of an impl
+        let (generics, generics_count) = if let Some(mut function_generics) = self.generics
+        {
+            let mut generics = Vec::new();
+            match object_type.clone() {
+                Type::Struct(_, mut struct_generics) => {
+                    generics.append(&mut struct_generics);
+                }
+                _ => {
+                    // TODO: Support other types that may contain generics such as `TraitAsType`
+                    unreachable!("Only struct method calls are supported");
+                }
+            };
+            let implicit_generics_count = generics.len();
+            generics.append(&mut function_generics);
+            (Some(generics), implicit_generics_count)
```

Comment:
Looking this over with https://github.com/noir-lang/noir/pull/5049 in mind, this snippet is incorrect I think - we shouldn't be implicitly adding struct generics here.

I believe you're doing this assuming that impls over a struct will always be generic over only the struct generics e.g.
```rs
impl<T> Foo<T> {
    ...
}
```
Since impl generics are automatically added to functions you're adding the struct generics here to try to match them. This is incorrect though since there are impls where this may not be the case, e.g:
```rs
impl Foo<Field> { ... }
```
It also relies on internal compiler details like the position that the impl generics are added in generic lists. This being off in some cases could be the reason you're getting some ambiguous type errors elsewhere since generic lists may be offset differently in e.g. trait impls.

I don't think we need to add struct generics implicitly at all. Since this PR was first made last year we've since added support for keeping track of a function's implicit generic count in the node interner already. We should be able to use that to skip over implicit generics if needed instead of trying to add them automatically here when we don't know what type should be used as the argument.

Reply:
I switched to just checking against the number of generics in `Forall` and subtracting the generic count from the function definition to get the implicit generic count. I instantiate any implicit generic as we do in the regular `instantiate` method rather than collecting any implicit generics. This provides a much simpler implementation.